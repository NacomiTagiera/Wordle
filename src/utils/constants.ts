import { LetterState } from '@/types';

export const NUMBER_OF_ROWS = 6;
export const NUMBER_OF_LETTERS = 5;

export const keyboardRows = [
  ['q', 'w', 'e', 'r', 't', 'y', 'u', 'i', 'o', 'p'],
  ['', 'a', 's', 'd', 'f', 'g', 'h', 'j', 'k', 'l', ''],
  ['enter', 'z', 'x', 'c', 'v', 'b', 'n', 'm', 'backspace'],
];

export const examples: {
  word: string;
  letterState: LetterState;
  description: string;
}[] = [
  {
    word: 'weary',
    letterState: 'correct',
    description: 'W is in the word and in the correct spot.',
  },
  {
    word: 'pills',
    letterState: 'present',
    description: 'I is in the word but in the wrong spot.',
  },
  {
    word: 'vague',
    letterState: 'absent',
    description: 'G is not in the word in any spot.',
  },
];
