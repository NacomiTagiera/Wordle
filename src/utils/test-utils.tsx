import React, { type PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { render, type RenderOptions } from '@testing-library/react';

import { type AppStore, type RootState, setupStore } from '@/redux/store';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
