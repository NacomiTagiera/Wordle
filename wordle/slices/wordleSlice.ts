import {
  bindActionCreators,
  createSlice,
  PayloadAction,
} from "@reduxjs/toolkit";

import { useAppDispatch, useAppSelector } from "@/app/hooks";
import { GameState } from "@/types";
import {
  appraiseGuess,
  getRandomWord,
  isGuessComplete,
  isValidWord,
  NUMBER_OF_LETTERS,
  NUMBER_OF_ROWS,
} from "@/lib/word-utils";

const initialState: GameState = {
  boardRowState: [],
  boardState: new Array(NUMBER_OF_ROWS).fill(""),
  currentRowIndex: 0,
  keyboardLetterState: {},
  message: { message: "", duration: 1500 },
  solution: getRandomWord(),
  status: "playing",
};

export const wordleSlice = createSlice({
  name: "wordle",
  initialState,
  reducers: {
    addLetter: (state, action: PayloadAction<string>) => {
      if (state.status !== "playing") return;
      if (state.boardState[state.currentRowIndex].length === NUMBER_OF_LETTERS)
        return;

      state.boardState[state.currentRowIndex] += action.payload;
    },
    removeLetter: (state) => {
      if (state.status !== "playing") return;
      if (!state.boardState[state.currentRowIndex].length) return;

      state.boardState[state.currentRowIndex] = state.boardState[
        state.currentRowIndex
      ].slice(0, -1);
    },
    resetGame: (state) => {
      state.boardRowState = [];
      state.boardState = new Array(NUMBER_OF_ROWS).fill("");
      state.currentRowIndex = 0;
      state.keyboardLetterState = {};
      state.message = { message: "", duration: 1500 };
      state.solution = getRandomWord();
      state.status = "playing";
    },
    resetMessage: (state) => {
      state.message = { message: "", duration: 1500 };
    },
    submitGuess: (state) => {
      const guess = state.boardState[state.currentRowIndex];

      if (!isGuessComplete(guess)) {
        state.message.message = "Not enough letters";
        return;
      }

      if (!isValidWord(guess)) {
        state.message.message = "Not in word list";
        return;
      }

      const boardRowState = appraiseGuess(guess, state.solution);

      state.boardRowState.push(boardRowState);

      boardRowState.forEach((letterState, index) => {
        const currentLetterState = state.keyboardLetterState[guess[index]];

        switch (currentLetterState) {
          case "present":
            if (letterState === "absent") break;

          case "correct":
            break;

          default:
            state.keyboardLetterState[guess[index]] = letterState;
            break;
        }
      });

      if (boardRowState.every((state) => state === "correct")) {
        state.status === "win";
        state.message = { message: "You win!", duration: 0 };
      }

      state.currentRowIndex++;

      if (state.currentRowIndex === state.boardState.length) {
        state.status = "fail";
        state.message = { message: state.solution, duration: 0 };
      }
    },
  },
});

export const useDispatchWordle = () => {
  const { actions } = wordleSlice;
  const dispatch = useAppDispatch();
  return bindActionCreators(actions, dispatch);
};

export const useGetBoardRows = () =>
  useAppSelector((state) =>
    state.boardState.map((letters, rowIndex) => ({
      letters,
      lettersState: state.boardRowState[rowIndex],
    }))
  );

export const useGetGameStatus = () => useAppSelector((state) => state.status);

export const useGetLetterState = () =>
  useAppSelector((state) => state.keyboardLetterState);

export const useGetMessage = () => useAppSelector((state) => state.message);

export default wordleSlice.reducer;
