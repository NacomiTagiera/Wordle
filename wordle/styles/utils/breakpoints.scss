$breakpoints-up: (
  "small": "415px",
  "medium": "768px",
  "large": "1024px",
);

@mixin breakpoint($size) {
  @media (min-width: map-get($breakpoints-up, $size)) {
    @content;
  }
}
