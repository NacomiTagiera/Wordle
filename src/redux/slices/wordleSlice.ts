import {
  bindActionCreators,
  createSelector,
  createSlice,
  type PayloadAction,
} from '@reduxjs/toolkit';

import { type GameState, type GameStatus, type LetterState } from '@/types';
import { NUMBER_OF_LETTERS, NUMBER_OF_ROWS } from '@/utils/constants';
import { appraiseGuess, getRandomWord, isGuessComplete, isValidWord } from '@/utils/word-utils';

import { useAppDispatch } from '../hooks';

const initialState: GameState = {
  boardRowState: [],
  boardState: new Array(NUMBER_OF_ROWS).fill('') as string[],
  currentRowIndex: 0,
  keyboardLetterState: {} as Record<string, LetterState>,
  message: { message: '', duration: 1500 },
  solution: getRandomWord(),
  status: 'playing' as GameStatus,
};

export const wordleSlice = createSlice({
  name: 'wordle',
  initialState,
  reducers: {
    addLetter: (state, action: PayloadAction<string>) => {
      if (state.status !== 'playing') return;
      if (state.boardState[state.currentRowIndex].length === NUMBER_OF_LETTERS) return;

      state.boardState[state.currentRowIndex] += action.payload;
    },
    removeLetter: (state) => {
      if (state.status !== 'playing') return;
      if (!state.boardState[state.currentRowIndex].length) return;

      state.boardState[state.currentRowIndex] = state.boardState[state.currentRowIndex].slice(
        0,
        -1
      );
    },
    resetGame: (state) => {
      state.boardRowState = [];
      state.boardState = new Array(NUMBER_OF_ROWS).fill('') as string[];
      state.currentRowIndex = 0;
      state.keyboardLetterState = {};
      state.message = { message: '', duration: 1500 };
      state.solution = getRandomWord();
      state.status = 'playing';
    },
    resetMessage: (state) => {
      state.message = { message: '', duration: 1500 };
    },
    submitGuess: (state) => {
      if (state.status !== 'playing') return;

      const guess = state.boardState[state.currentRowIndex];

      if (!isGuessComplete(guess)) {
        state.message.message = 'Not enough letters';
        return;
      }

      if (!isValidWord(guess)) {
        state.message.message = 'Not in word list';
        return;
      }

      const boardRowState = appraiseGuess(guess, state.solution);

      state.boardRowState.push(boardRowState);

      boardRowState.forEach((letterState, index) => {
        const currentLetterState = state.keyboardLetterState[guess[index]];

        if (currentLetterState !== 'present' && currentLetterState !== 'correct') {
          state.keyboardLetterState[guess[index]] = letterState;
        }
      });

      if (state.currentRowIndex === state.boardState.length - 1) {
        state.status = 'fail';
        state.message = { message: state.solution, duration: 0 };
      }

      if (boardRowState.every((state) => state === 'correct')) {
        state.status = 'win';
        state.message = { message: 'You win!', duration: 0 };
      }

      state.currentRowIndex++;
    },
  },
});

export const { addLetter, removeLetter, resetGame, resetMessage, submitGuess } =
  wordleSlice.actions; //for tests

export const useDispatchWordle = () => {
  const { actions } = wordleSlice;
  const dispatch = useAppDispatch();
  return bindActionCreators(actions, dispatch);
};

const selectGameState = (state: GameState) => state;

export const selectBoardRows = createSelector(selectGameState, (state) =>
  state.boardState.map((letters, rowIndex) => ({
    letters,
    lettersState: state.boardRowState[rowIndex],
  }))
);

export const selectGameStatus = (state: GameState) => state.status;

export const selectLetterState = (state: GameState) => state.keyboardLetterState;

export const selectMessage = (state: GameState) => state.message;

export const { reducer } = wordleSlice;
