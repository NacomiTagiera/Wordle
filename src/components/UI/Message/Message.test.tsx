import { screen, waitFor } from '@testing-library/react';

import { initialState } from '@/redux/slices/wordleSlice';
import { renderWithProviders } from '@/utils/test-utils';

import { Message } from '.';

describe('Message', () => {
  it('matches the snapshot', () => {
    const { asFragment } = renderWithProviders(<Message />, {
      preloadedState: {
        wordle: {
          ...initialState,
          message: {
            message: 'Test',
            duration: 500,
          },
        },
      },
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('displays provided message', () => {
    renderWithProviders(<Message />, {
      preloadedState: {
        wordle: {
          ...initialState,
          message: {
            message: 'Test',
            duration: 500,
          },
        },
      },
    });

    const message = screen.getByTestId('message');
    expect(message).toBeVisible();
    expect(message).toHaveTextContent('Test');
  });

  it('closes automatically after some time', async () => {
    const { store } = renderWithProviders(<Message />, {
      preloadedState: {
        wordle: {
          ...initialState,
          message: {
            message: 'Test',
            duration: 1,
          },
        },
      },
    });

    screen.getByText('Test');
    await waitFor(() => expect(store.getState().wordle.message.message).toEqual(''), {
      timeout: 1,
    });
    await waitFor(() => expect(screen.queryByText('Test')).toBeNull());
  });

  it('does not render when there is no message', () => {
    renderWithProviders(<Message />);

    expect(screen.queryByTestId('message')).not.toBeInTheDocument();
  });
});
