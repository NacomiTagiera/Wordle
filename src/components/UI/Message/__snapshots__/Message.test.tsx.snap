// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiSnackbar-root MuiSnackbar-anchorOriginTopCenter css-1q8jzsy-MuiSnackbar-root"
    data-testid="message-container"
    role="presentation"
  >
    <p
      class="message"
      direction="down"
      style="opacity: 1; transform: scale(1, 1); padding: 13px; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      Test
    </p>
  </div>
</DocumentFragment>
`;
